src/box/memtx_tx.c

```diff
@@ -1622,10 +1622,16 @@ memtx_tx_story_insert_is_visible(struct memtx_story *story, struct txn *txn,
 {
 	*is_own_change = false;
 
-	if (story->add_stmt != NULL && story->add_stmt->txn == txn) {
-		/* Tuple is added by us (@txn). */
-		*is_own_change = true;
-		return true;
+	if (story->add_stmt != NULL) {
+		struct txn *add_stmt_txn = story->add_stmt->txn;
+		if (add_stmt_txn == txn) {
+			/* Tuple is added by us (@txn). */
+			*is_own_change = true;
+			return true;
+		} else if (txn_has_flag(add_stmt_txn, TXN_EARLY_ACK) &&
+			   add_stmt_txn->signature >= 0) {
+			return true; /* Tuple addition was acked to the user. */
```

Comment:
Why does the txn still live in any tx managers? You only need its rows to be able to undo it. It doesn't need to stay in any registries, managers, stories, etc. We only need its data after it was written WAL and acked. Nothing else. What do I miss?

Reply:
Nope, at least the memtx MVCC transaction manager needs it:
https://github.com/tarantool/tarantool/blob/17b11601a45060523120fb6102e02b1a30332fbe/src/box/memtx_engine.cc#L674-L675

> You only need its rows to be able to undo it

We don't have an `undo` mechanism in the storage engine, we only have `commit` or `rollback`. A transaction can either have a committed state, after which it cannot be rolled back (undone), or a non-committed state, during which it can be rolled back (undone).

Reply:
Yeah, you have good points.

> for instance, we want there on_commit or on_rollback triggers to get executed only when they are confirmed or rolled back synchronously, not when a success is returned to the user.

Hm? When a success is returned, the changes are already visible. We can't not invoke `on_commit` triggers here. If a txn is acked to the user, it is fully committed. If it is going to be rolled back later, that would be another txn. Postponing `on_commit` is dangerous if such a trigger in user's code was supposed to update some global states before anybody accesses the newly visible changes. For instance, vshard uses that for _bucket space updates. `On_commit` needs to immediately update the most actual in-memory states.

> Secondly, AFAIC, we currently don't have any undo statement mechanism in the storage engines, 

You don't need it, hm? Txn change-list contains statements, each having old and new tuple and the space. "Undo" means walk the list backwards delete "new" and insert "old". We don't even need `struct txn` for it. Only its statements and its memory region (to avoid copying the statements), really. It is just convenient to store the txn in all limbo entries. But some of them no problem could be literally fully finished and kept only for their undo data.

> A transaction can either have a committed state, after which it cannot be rolled back (undone), or a non-committed state, during which it can be rolled back (undone).

Yes. If acked, it means committed. The txn itself is done. Can't remain in any engines, conflict managers, etc. Its "rollback" after ack is already another txn, not the same one. Internal txn, but a totally new one anyway. Wouldn't it simplify things a lot? What are the problems with it?

Reply:
Interesting. Imagine there is an async txn which was early-acked. If there were more transactions after then the following is possible:
- Sync txns are done and committed. It means this async txn is fully confirmed, won't be rolled back.
- Sync txns were created later, but not committed yet. Then all is rolled back. Those sync txns were not even visible in mvcc. And without mvcc they are going to be rolled back in reversed order same as I suggested. But automatically, by txn_rollback or whatever.
- Async txns were created later. They are hanging the same way then. And rollback will go in the reversed order like I described.

I.e. if we have an async early-acked txn in the limbo, then all the next txns are also early-acked or not committed at all, and they are all in the limbo too. Hence this way of rollback must work fine. Do I miss something important? Can easily happen.

Reply:
> > for instance, we want there on_commit or on_rollback triggers to get executed only when they are confirmed or rolled back synchronously, not when a success is returned to the user.

> Hm? When a success is returned, the changes are already visible. We can't not invoke on_commit triggers here. If a txn is acked to the user, it is fully committed. If it is going to be rolled back later, that would be another txn. Postponing on_commit is dangerous if such a trigger in user's code was supposed to update some global states before anybody accesses the newly visible changes. For instance, vshard uses that for _bucket space updates. On_commit needs to immediately update the most actual in-memory states.

Agree. "early ack" is the point where we commit the transaction. So the on_commit triggers should be fired there.

>> Secondly, AFAIC, we currently don't have any undo statement mechanism in the storage engines,

> You don't need it, hm? Txn change-list contains statements, each having old and new tuple and the space. "Undo" means walk the list backwards delete "new" and insert "old". We don't even need struct txn for it. Only its statements and its memory region (to avoid copying the statements), really. It is just convenient to store the txn in all limbo entries. But some of them no problem could be literally fully finished and kept only for their undo data.

That's a good point. I just haven't thought we could store committed transactions there. Let's go this way, sure.

Reply:
How are we going to tolerate `undo` failures in the case of vinyl? Panic?

Reply:
@locker do you have any thoughts on this?

Reply:
Good questions! Can't answer about Vinyl or memcs, simply don't know them. But seems like it shouldn't depend on any engine. The undo data is stored in `txn.stmts` list, regardless of the engine. Why do you care about the engine? Maybe you said before, but I forgot, sorry.

> I think that the main problem with this approach is that it breaks the transaction abstraction.

Would you define what is "transaction abstraction"? AFAIS, a txn gets created and committed / rolled back once. It can't be committed and then rolled back. It is not possible. It can't be the same transaction. If it got committed, literally in WAL marked as committed, then it is done. You have to fire the commit triggers. Rolling back is not possible. You can undo the data, but not literally undo that same WAL-committed txn.

When you define a txn as a bunch of changes with a unique ID which gets committed or rolled back only once, then the definition seems to match with the proposed solution. Undo the changes, but in another txn.

Reply:
> The undo data is stored in `txn.stmts` list, regardless of the engine.

That's not correct, at least our engine implementations do not follow that:

1. vinyl does not store the undo data in `txn.stmts` — it has a separate transaction view for that, which gets destroyed when the transaction gets committed by the engine. As I said, it does not set the `stmt->old_tuple`, for instance (for obvious performance reasons): https://github.com/tarantool/tarantool/blob/2300704e8317f2d8a545cde1394f8cbbb7e95741/src/box/vinyl.c#L2302-L2313
2. memcs also does not store the undo data in `txn.stmts` — it has a separate `undo` structure for that, which gets destroyed when the transaction gets committed by the engine. It also does not set the `stmt->old_tuple`: https://github.com/tarantool/tarantool-ee/blob/0aeed64402e9408907daaa5d453f49acd82e5470/src/box/memcs_engine.c#L105-L113

Reply:
> If it got committed, literally in WAL marked as committed, then it is done.

This is also not correct. In the synchronous replication context we do not mark transactions as committed in the WAL — the commit/rollback boundaries are defined by `PROMOTE`/`DEMOTE`/`CONFIRM`/`ROLLBACK` entries.

Reply:
Ok, good point about the txn.stmts, I didn't know the other engines don't keep it up to date, which is btw puzzling. Because the public on-commit and on-rollback triggers use txn.stmts, not engine-specific data. Are your sure in vinyl it is not kept up to date? Can you make a test with on-commit triggers which shows that?

> AFAIC, it is possible to fire the commit triggers and make the transaction visible as committed, but not commit the transaction in the engines.

That is an implementation detail. "Rollback" of the txn still would have to be a new txn undoing the changes of the first txn.

I still don't see why do we need the engine txns kept not closed. Visible but not committed? So there is undo data somewhere anyway, right? Why not store it in txn.stmts then?

Speaking of Vinyl performance - the on-commit triggers are public. If they work incorrectly right now due to undo-data being internal, then we need to fix it. And that would make undo-data available at least when the triggers are installed. Secondly, we could make it also enforced when this weird async-sync mode works. We need undo data. There is no choice. To undo a vinyl-replace you need the old tuple. What am I missing, help?

> When we replay the logs, we are supposed to roll back the rejected transactions

You are speaking of synchro txns here. Yes, we can, because they were not committed. Synchro txns are committed when get CONFIRM in WAL. Not when they just get into WAL. After a synchro txn gets CONFIRM it is never rolled back. Even by PROMOTE/DEMOTE.

> And you propose to write a new transaction during this process.

No. We don't do that now for synchro txns and it won't be done. Because they are not committed until CONFIRM. And after CONFIRM yes, they can't be rolled back even now. For async txns - yes. For them you need to write their undo data into WAL. And you can't use same txn ID already for that. Which means it is a rollback in a new txn. Is it?

> We cannot write a new transaction every time 

For undoing the already committed async txns I see no other way. We either need to make all txns sync when synchro is enabled, or we need to undo async txns as separate system-created txns. Their undo data needs to be in WAL. And it can't be same txn ID. The relation to synchro txns here seems irrelevant, or? They remain working like they do now, because they are not and won't be undone by the system after being committed (their "commit" moment is CONFIRM WAL write, not the changeset WAL write).

Synchro txns don't have this problem, because they get explicit commit via CONFIRM. Until then they can be rolled back without having their undo data written into WAL.

Async txns after reaching WAL are committed and visible. From WAL PoV those txns are done, finished, have no continuation. To "rollback" them you need to write their undo data into WAL. Doing it with the same txn ID and timestamp is going to be a bug, because txn ID and timestamp in WAL won't be either monotonic or unique. By all means it is a new txn. I feel like I miss something important which somehow isn't obvious to me but visible to the others?